aptos-move/aptos-vm/src/gas.rs

```diff
@@ -113,44 +114,58 @@ pub fn get_gas_parameters(
+///
+/// The current setup consists of the standard gas meter & algebra + the memory usage tracker.
+pub fn make_prod_gas_meter(
+    gas_feature_version: u64,
+    vm_gas_params: VMGasParameters,
+    storage_gas_params: StorageGasParameters,
+    is_governance_proposal: bool,
+    meter_balance: Gas,
+) -> ProdGasMeter {
+    MemoryTrackedGasMeter::new(StandardGasMeter::new(StandardGasAlgebra::new(
+        gas_feature_version,
+        vm_gas_params,
+        storage_gas_params,
+        is_governance_proposal,
+        meter_balance,
+    )))
+}
+
 pub(crate) fn check_gas(
     gas_params: &AptosGasParameters,
     gas_feature_version: u64,
     resolver: &impl AptosMoveResolver,
     txn_metadata: &TransactionMetadata,
     features: &Features,
+    is_governance_proposal: bool,
     log_context: &AdapterLogSchema,
 ) -> Result<(), VMStatus> {
     let txn_gas_params = &gas_params.vm.txn;
     let raw_bytes_len = txn_metadata.transaction_size;
     // The transaction is too large.
     if txn_metadata.transaction_size > txn_gas_params.max_transaction_size_in_bytes {
-        let data =
-            resolver.get_resource(&CORE_CODE_ADDRESS, &ApprovedExecutionHashes::struct_tag());
-
-        let valid = if let Ok(Some(data)) = data {
-            let approved_execution_hashes = bcs::from_bytes::<ApprovedExecutionHashes>(&data).ok();
-            let valid = approved_execution_hashes
-                .map(|aeh| {
-                    aeh.entries
-                        .into_iter()
-                        .any(|(_, hash)| hash == txn_metadata.script_hash)
-                })
-                .unwrap_or(false);
-            valid
-                // If it is valid ensure that it is only the approved payload that exceeds the
-                // maximum. The (unknown) user input should be restricted to the original
-                // maximum transaction size.
-                && (txn_metadata.script_size + txn_gas_params.max_transaction_size_in_bytes
-                >= txn_metadata.transaction_size)
-                // Since an approved transaction can be sent by anyone, the system is safer by
-                // enforcing an upper limit on governance transactions just so something really
-                // bad doesn't happen.
-                && txn_metadata.transaction_size <= MAXIMUM_APPROVED_TRANSACTION_SIZE.into()
+        let max_txn_size_gov = if gas_feature_version >= RELEASE_V1_13 {
+            gas_params.vm.txn.max_transaction_size_in_bytes_gov
         } else {
-            false
+            MAXIMUM_APPROVED_TRANSACTION_SIZE_LEGACY.into()
         };
 
+        let valid = is_governance_proposal
+            // If it is valid ensure that it is only the approved payload that exceeds the
+            // maximum. The (unknown) user input should be restricted to the original
+            // maximum transaction size.
+            && (txn_metadata.script_size + txn_gas_params.max_transaction_size_in_bytes
```

Comment:
I see now. I think the flow is confusing: can we instead do 1) check if is governance, then check using custom limits, else 2) check script size + normal limits? Fitting everything into a single statement is very confusing

Reply:
Refactored this piece of code as you suggested. Agreed that the original logic was quite confusing and it would be better to branch based on the whether the txn is a governance script or not at the top.